### PR TITLE
feat(dotnet): add license support for NuGet

### LINKED ISSUE #5216: feat(NuGet): add license support
## Description
`NuGet` has [*.nuspec](https://learn.microsoft.com/en-us/nuget/reference/nuspec) files with information about licenses.
We can use this for detect licenses.

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
`packages.config` and `packages.lock.json` files don't have information about the licenses used.

Add support of [*.nuspec](https://learn.microsoft.com/en-us/nuget/reference/nuspec) files from [global packages folder](https://learn.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders).

notes:
- Currently only the default path and `NUGET_PACKAGES` environment variable are supported.
- `licenseUrl` is not supported, because this field has been deprecated.
- license `expression` only supports

Example:
```
➜  ./trivy fs -f json --list-all-pkgs ./packages.lock.json 
2023-09-20T13:50:05.734+0600	INFO	Number of language-specific files: 1
2023-09-20T13:50:05.734+0600	INFO	Detecting nuget vulnerabilities...
2023-09-20T13:50:05.734+0600	DEBUG	Detecting library vulnerabilities, type: nuget, path: packages.lock.json
{
  ...
  "Results": [
    {
      "Target": "packages.lock.json",
      "Class": "lang-pkgs",
      "Type": "nuget",
      "Packages": [
        {
          "ID": "Newtonsoft.Json@13.0.3",
          "Name": "Newtonsoft.Json",
          "Version": "13.0.3",
          "Licenses": [
            "MIT"
          ],
          "Layer": {},
          "Locations": [
            {
              "StartLine": 5,
              "EndLine": 10
            }
          ]
        }
      ]
    }
  ]
}
```

## Related issues
- Close #5216

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		return nil, xerrors.Errorf("unable to open %q file: %w", path, err)
+	}
+	defer func() { _ = f.Close() }()
+
+	var pkg Package
+	if err = xml.NewDecoder(f).Decode(&pkg); err != nil {
+		return nil, xerrors.Errorf("unable to decode %q file: %w", path, err)
+	}
+
+	if license := pkg.Metadata.License; license.Type == "expression" && license.Text != "" {
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go:78]: [SPDX license identifier](https://spdx.org/licenses/) is uses for `expression` license type. For custom licenses  the `file` type is uses - https://learn.microsoft.com/en-us/nuget/reference/nuspec#license
SPDX includes most popular licenses. I think we can skip parse of `file` type for now.
If we receive requests for this - we will be able to add  file parsing using [licenseclassifier](https://github.com/google/licenseclassifier) later.
@knqyf263 wdyt?

Comment (knqyf263) [reply pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go:78]: Agree. Let's go for the small start.

File: pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		return []string{license.Text}, nil
+	}
+
+	return nil, nil
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go:82]: nit: I prefer the happy path getting aligned to the left edge.
https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

```suggestion
	if license := pkg.Metadata.License; license.Type != "expression" || license.Text == "" {
		return nil, nil
	}

	return []string{pkg.Metadata.License}, nil
```

File: pkg/fanal/analyzer/language/dotnet/nuget/nuget.go [hunk 3]
```diff
@@ -2,59 +2,112 @@ package nuget
...
+
+		for i, lib := range app.Libraries {
+			license, ok := foundLicenses[lib.ID]
+			if !ok {
+				license, err = a.licenseParser.findLicense(lib.Name, lib.Version)
+				if err != nil {
+					if !errors.Is(err, fs.ErrNotExist) {
+						return xerrors.Errorf("license find error: %w", err)
+					}
+				}
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/dotnet/nuget/nuget.go:82]: nit: We might want to aggregate the statements.

```suggestion
				if err != nil && !errors.Is(err, fs.ErrNotExist) {
					return xerrors.Errorf("license find error: %w", err)
				}
```

**Review by knqyf263 — COMMENTED** [review]
Looks very nice! I left small comments.

File: pkg/fanal/analyzer/language/dotnet/nuget/nuget.go (hunk 3 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dotnet/nuget/nuget.go]: Right! Thanks! Updated in https://github.com/aquasecurity/trivy/pull/5217/commits/1d0fed1336f1501c5d1f3d1b453be7fc47fa1ba9.

File: pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go (hunk 2 above)
Comment (DmitriyLewen) [reply pkg/fanal/analyzer/language/dotnet/nuget/nuspec.go]: Changed in https://github.com/aquasecurity/trivy/pull/5217/commits/1d0fed1336f1501c5d1f3d1b453be7fc47fa1ba9

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
CLA assistance didn't work. I opened the following link to trigger it.
https://cla-assistant.io/check/aquasecurity/trivy?pullRequest=5217

**Pull request merged**